File: backend/src/main/java/org/certificaciones/funproeibbackend/service/DocumentoService.java
package org.certificaciones.funproeibbackend.service;

import org.certificaciones.funproeibbackend.dto.ArchivoDescargable;
import org.certificaciones.funproeibbackend.dto.DocumentoRequest;
import org.certificaciones.funproeibbackend.dto.DocumentoResponse;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

public interface DocumentoService {
    DocumentoResponse registrar(DocumentoRequest request);
    DocumentoResponse subirArchivo(Long idPostulacion, Long idReqDocumento, MultipartFile archivo);
    ArchivoDescargable descargar(Long idDocumento);
    List<DocumentoResponse> listarPorPostulacion(Long idPostulacion);
    DocumentoResponse verificar(Long idDocumento);
    void verificarTodosDePostulacion(Long idPostulacion);
}

File: frontend/src/app/core/services/documento.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ApiService } from './api.service';
import { DocumentoRequest, DocumentoResponse, ReqDocumentoResponse } from '../models/documento.model';

@Injectable({ providedIn: 'root' })
export class DocumentoService {
  constructor(private api: ApiService, private http: HttpClient) {}

  listarRequisitos(idPrograma: number): Observable<ReqDocumentoResponse[]> {
    return this.api.get<ReqDocumentoResponse[]>(`/req-documentos/programa/${idPrograma}`);
  }

  listarPorPostulacion(idPostulacion: number): Observable<DocumentoResponse[]> {
    return this.api.get<DocumentoResponse[]>(`/documentos/postulacion/${idPostulacion}`);
  }

  registrar(request: DocumentoRequest): Observable<DocumentoResponse> {
    return this.api.post<DocumentoResponse>('/documentos', request);
  }

  subirArchivo(idPostulacion: number, idReqDocumento: number, archivo: File): Observable<DocumentoResponse> {
    const formData = new FormData();
    formData.append('idPostulacion', String(idPostulacion));
    formData.append('idReqDocumento', String(idReqDocumento));
    formData.append('archivo', archivo);
    return this.http.post<DocumentoResponse>(`${this.api.base}/documentos/subir`, formData);
  }

  /** URL para ver/descargar el archivo de un documento cargado. */
  urlArchivo(idDocumento: number): string {
    return `${this.api.base}/documentos/${idDocumento}/archivo`;
  }

  verificar(idDocumento: number): Observable<DocumentoResponse> {
    return this.api.patch<DocumentoResponse>(`/documentos/${idDocumento}/verificar`);
  }
}
